### DISCUSSION TITLE
Korean input - maybe a bug

### DISCUSSION THREAD

**Discussion by Dhghomon** [emilk/egui#414] [Q&A]
I have another interesting discussion a bit similar to the double paste bug that I saw was recently fixed. (I thought it had something to do with me adding two new fonts but then I came across the release notes on this)

So this has to do with Korean input on phones. First of all, here's the site: https://www.hanjabuddy.com/ You paste or type Korean into there and then it looks at the position of the cursor and finds matching hanja (Chinese characters in Korean) which show up as buttons. You then click on the button you want and it turns the hangul into that hanja.

Now on iPhones it's adding characters to the string as soon as a button is pressed. But Korean characters take a few keystrokes to finish, so it's adding these unfinished characters before they're done. For example, let's say I want to write ''han''. It's ㅎ (h), then ㅏ (a), and ㄴ (n). So you do this:

Type the g key: it's now ㅎ.
Type the k key: it's now 하. This is actually a finished character ''ha'', which the user might want to type. But it doesn't move to the next space yet because the user might want to add another consonant.
Type the s key: it's now 한. ''han'' is now done, but even now it is still waiting for extra input because you could still fit in an extra something in the lower right. You could make it 핝 핞 for example (not real words, but technically writable)

This image was sent to me by a friend who gave it a try on her iPhone where she tried to type 아 심심하다 (Ah I'm bored). It ended up like this:

ㅇ아 ㅅ시심ㅅ시심ㅎ하핟하다

![image](https://user-images.githubusercontent.com/56599343/118779711-4bd1ad00-b8c6-11eb-9916-65b48535a40d.png)

So it's taking every step of the character being created and adding it to the string to paste in. Here are the characters being made:

ㅇ -> 아
ㅅ -> 시 -> 심
ㅎ -> 하 -> 핟 -> 하다

The last one is interesting because it's 하다 (ha da). In the middle where she's typed up to ''had'' it's still on the first character, but once an a is added it becomes clear that it's ha and da, and the d (ㄷ) moves over to become the first part of the next character.

But beyond all the Korean background, the fix might be to avoid pasting in until the cursor has moved.

I'm loving egui by the way! It's the first time I've been able to compile to WASM thanks to the scripts you have that move things to the /doc/ folder. I experimented a bit before with WASM but got annoyed with trying to get the bindings working and didn't really have a reason to want to compile to it in any case and just put it off until later.

**Comment by emilk** [comment]
Cool app!

I have no experience with IME. It was implemented by @n2 in https://github.com/emilk/egui/pull/253

**Reply by Dhghomon** [reply]
Oh yeah, that looks like the place. I guess for iOS users it's giving `true` for the `!text.is_empty()` condition on line 755 for every time they make a keystroke even before the character is done composing.

If anyone wants to try it in practice, there's a popup keyboard here:

https://translate.google.com/?sl=ko&tl=en&op=translate

Once it's popped up, typing gkek will show 하다 in the following four stages:

Type g-> ㅎ (incomplete character)
Type k -> 하 (하 is a complete character, but it's still possible that the user will want to add more so it's not done yet)
Type e -> 핟 (shows up as 핟 now, which is technically composable but a nonsense word.)
Type k -> 하다 (here the computer sees that the preceding ㄷ and ㅏ are meant to be a separate character and splits them into two)

Typing this all out makes me realize just how complicated it must have been to implement Korean input for the first time way back when. I guess an equivalent to English would be like this:

User has typed *sleep* -> it's a real word but the user could still be intending to type *sleepy*, *sleepiness*, etc.

One option: user types spacebar -> the computer sees that the user has typed the word "sleep" and registers this as input.

Another option: user types y -> now the word is *sleepy*, and no other possible words can happen so the word sleepy gets registered as the input even without hitting the spacebar.

**Comment by n2** [comment]
The behavior of the Korean input method may be different, and my event logic is insufficient. Let me figure it out.

**Reply by Dhghomon** [reply]
@n2 Thank you! Much appreciated and let me know if you have any questions about Korean input.

Really funny that it only does this in iOS though. On a PC it's flawless.

**Reply by he4rtbleed** [reply]
Did you solve this problem?
I have exactly same issue now.
